fashion-messaging-app-api/server.js

```diff
@@ -173,45 +172,103 @@ app.get('/recommendations', async (req, res) => {
-      where: { userId: req.session.user.id },
+      where: { 
+        userId: req.session.user.id,
+        searchTerm: {
+          [Op.not]: defaultQuery 
+        }
+      },
       order: [['createdAt', 'DESC']],
     });
 
-    const mostFrequentSearch = await SearchResult.findOne({
-      where: { userId: req.session.user.id },
-      order: [['searchCount', 'DESC']],
+    const popularSearch = await SearchResult.findOne({
+      attributes: ['searchTerm', [sequelize.fn('COUNT', sequelize.col('searchTerm')), 'searchCount']],
+      group: ['searchTerm'],
+      where: {
+        searchTerm: {
+          [Op.not]: defaultQuery  
+        }
+      },
+      order: [[sequelize.fn('COUNT', sequelize.col('searchTerm')), 'DESC']],
+      raw: true,
     });
 
-    let lastSearchPhotos = [];
-    let mostFrequentSearchPhotos = [];
+    const latestPost = await Post.findOne({
+      where: { 
+        userId: req.session.user.id
+      },
+      order: [['createdAt', 'DESC']],
+    });
 
-    if (lastSearch) {
-      const response = await clientAPI.photos.search({ 
-        query: lastSearch.searchTerm, 
+
+    if (latestPost) {
+      const responseLatestPost = await clientAPI.photos.search({ 
+        query: latestPost.title,  
         per_page: PHOTOS 
       });
-      lastSearchPhotos = response.photos;
 
+      responseLatestPost.photos.forEach(photo => {
+        photo.source = "latest post";
+      });
+    
+      recommendations.push(...responseLatestPost.photos);
+    }
+
+   
+    if (!frequentSearch && !lastSearch && !popularSearch) {
+      return res.json([]);  
     }
 
-    if (mostFrequentSearch) {
-      const response = await clientAPI.photos.search({ 
-        query: mostFrequentSearch.searchTerm, 
+    
+    if (frequentSearch) {
+      const responseFrequentSearch = await clientAPI.photos.search({ 
+        query: frequentSearch.searchTerm, 
         per_page: PHOTOS 
       });
-      mostFrequentSearchPhotos = response.photos;
+     
+      responseFrequentSearch.photos.forEach(photo => {
+        photo.source = "most frequent search";
+      });
+      recommendations.push(...responseFrequentSearch.photos);
     }
 
-let photos = [...lastSearchPhotos, ...mostFrequentSearchPhotos];
-photos = [...new Map(photos.map(photo => [photo.id, photo])).values()];
-
+    
+    if (lastSearch && (!frequentSearch || (frequentSearch && frequentSearch.searchTerm !== lastSearch.searchTerm))) {
+      const responseLastSearch = await clientAPI.photos.search({ 
+        query: lastSearch.searchTerm, 
+        per_page: PHOTOS 
+      });
+      
+      responseLastSearch.photos.forEach(photo => {
+        photo.source = "last search";
+      });
+      recommendations.push(...responseLastSearch.photos);
+    }
+    if (popularSearch && (!frequentSearch || (frequentSearch && frequentSearch.searchTerm !== popularSearch.searchTerm)) && (!lastSearch || (lastSearch && lastSearch.searchTerm !== popularSearch.searchTerm))) {
```

Comment:
We usually try to avoid having so many ||, &&, and ! checks at once because it becomes super hard to read. You can break this up by defining variables or creating function(s)

Reply:
> We usually try to avoid having so many ||, &&, and ! checks at once because it becomes super hard to read. You can break this up by defining variables or creating function(s)

Okay, got it!